### PR TITLE
[WIP] Add HTTP request profiling.

### PR DESCRIPTION
**Pull request by petrabarus**
Just putting this WIP PR here. I am still trying to think a good design for this. The one that will minimize exposure to AWS S3 yet still have complete functions.

### CODE REVIEW

File: src/main/java/tdl/s3/sync/Destination.java [hunk 1]
```diff
@@ -41,10 +57,67 @@ public AWSSecretsProvider getSecret() {
         return secret;
     }
 
-    public AmazonS3 buildClient() {
-        return AmazonS3ClientBuilder.standard()
+    public AmazonS3 getClient() {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/sync/Destination.java]: At the end of this refactoring, there should be no need for getClient(), getBucket() and getPrefix()

File: src/main/java/tdl/s3/sync/Destination.java [hunk 2]
```diff
@@ -41,10 +57,67 @@ public AWSSecretsProvider getSecret() {
...
+
+    public ObjectMetadata getObjectMetadata(String bucket, String key) {
+        return this.client.getObjectMetadata(bucket, key);
+    }
+    
+    public String getFullPath(String path) {
+        return getPrefix()+ path;
+    }
+
+    public boolean canUpload(String path) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/sync/Destination.java]: I would rename `path` to `remotePath`

File: src/main/java/tdl/s3/sync/Destination.java [hunk 3]
```diff
@@ -41,10 +57,67 @@ public AWSSecretsProvider getSecret() {
...
+    }
+
+    private void buildClient() {
+        this.client = AmazonS3ClientBuilder.standard()
                 .withCredentials(secret)
                 .withRegion(secret.getS3Region())
                 .build();
     }
+
+    public ObjectMetadata getObjectMetadata(String bucket, String key) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/sync/Destination.java]: This is not used any more and can be removed.

File: src/main/java/tdl/s3/sync/Destination.java [hunk 4]
```diff
@@ -41,10 +57,67 @@ public AWSSecretsProvider getSecret() {
...
+        } catch (AmazonS3Exception ex) {
+            if (ex.getStatusCode() == 404) {
+                return false;
+            } else {
+                throw ex;
+            }
+        }
+    }
+    
+    public String initUploading(String remotePath) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/s3/sync/Destination.java]: I would name the returned String by creating a simple type UploadId

File: src/main/java/tdl/s3/sync/Destination.java (hunk 1 above)
Comment (petrabarus) [inline src/main/java/tdl/s3/sync/Destination.java]: `getClient()` is required to mock the S3 client call in the `UnfinishedWritingFileUploadingStrategyTest`.
The `getBucket()` and `getPrefix()` just shorthands the `secrets`.

The `MultipartUploadHelper` class needs the bucket and prefix. For example method `getAlreadyUploadedParts` and `getPartListing`. How do you suggest we hide it inside the `Destination` class?

Comment (julianghionoiu) [inline src/main/java/tdl/s3/sync/Destination.java]: `UnfinishedWritingFileUploadingStrategyTest` can be removed for now. It is preventing us from changing the code. We will focus on the code coverage metric in a future iteration.

`getAlreadyUploadedParts` and `getPartListing` can move inside `Destination`

Comment (petrabarus) [inline src/main/java/tdl/s3/sync/Destination.java]: Okay, got it. Thanks!

Comment (petrabarus) [inline src/main/java/tdl/s3/sync/Destination.java]: In the https://github.com/julianghionoiu/s3-sync-stream/pull/18/commits/f1ffa69968de843d8e4b75124dfe8920f9ccdf02 the `getBucket`, `getPrefix`, and `getClient` are removed.

I guess this is the best refactoring I can get without making `Destination` class really overbloated.

**Pull request merged**